### PR TITLE
ISPN-15390 Allow to project the query score

### PR DESCRIPTION
**Pull request by fax4ever**
https://issues.redhat.com/browse/ISPN-15390

~~Draft still to test/address: broadcast queries, remote queries.~~ done!

### CODE REVIEW

**Comment by fax4ever** [comment]
@yrodiere @marko-bekhta I don't if you want to have a look at it

File: query/src/test/java/org/infinispan/query/projection/ScoreProjectionTest.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+   @Test
+   public void nativeQuery() {
+      ComponentRegistry componentRegistry = ComponentRegistry.of(cache);
+      SearchMapping searchMapping = componentRegistry.getComponent(SearchMapping.class);
+      assertThat(searchMapping).isNotNull();
+
+      SearchSession mappingSession = searchMapping.getMappingSession();
+      SearchScope<Game> scope = mappingSession.scope(Game.class);
+
+      SearchProjection<List<?>> projection = scope.projection().composite(scope.projection().score(), scope.projection().entity()).toProjection();
```
Comment (marko-bekhta) [inline query/src/test/java/org/infinispan/query/projection/ScoreProjectionTest.java:94]: Just wanted to share that not so long ago, Hibernate Search introduced a way to bind this type of projections like:
```java
@ProjectionConstructor
public record MyProjection(@ScoreProjection float score, @EntityProjection Game entity) {
}
// and then
SearchProjection<MyProjection> projection = scope.projection().composite().as( MyProjection.class ).toProjection();
```
Not that it is needed here, but maybe you'll find it useful in some future tests 😃 
See https://docs.jboss.org/hibernate/stable/search/reference/en-US/html_single/#search-dsl-projection-entity-mapping

**Review by marko-bekhta — COMMENTED** [review]
I didn't find aa lot of Hibernate Search-related things in this PR so cannot add much, but looks nice overall 😃

**Comment by karesti** [comment]
@fax4ever I'm going to build this PR and test it in langchain SPI

File: query/src/test/java/org/infinispan/query/projection/ScoreProjectionTest.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+   @Test
+   public void entityScoreAndField() {
+      Query<Object[]> query;
+      List<Object[]> games;
+
+      query = cache.query("select g.name, score(g), g from org.infinispan.query.model.Game g where g.description : 'description'~2");
+      games = query.list();
+      assertThat(games).extracting(objects -> objects[0]).containsExactly("1", "2", "3", "4");
+      assertThat(games).extracting(objects -> objects[1]).hasOnlyElementsOfType(Float.class).isNotNull();
```
Comment (karesti) [inline query/src/test/java/org/infinispan/query/projection/ScoreProjectionTest.java:81]: in langchain, the score is a double

File: query/src/test/java/org/infinispan/query/projection/ScoreProjectionTest.java (hunk 1 above)
Comment (fax4ever) [reply query/src/test/java/org/infinispan/query/projection/ScoreProjectionTest.java:94]: Thank you @marko-bekhta nice API. Thanks for sharing it!

File: query/src/main/java/org/infinispan/query/dsl/embedded/impl/SearchQueryBuilder.java [hunk 3]
```diff
@@ -72,15 +71,12 @@ public LuceneSearchQuery<List<Object>> keyAndEntity() {
       SearchProjectionFactory<EntityReference, ?> projectionFactory = scope.projection();
       SearchProjection<?>[] searchProjections = new SearchProjection<?>[]{
             projectionFactory.entityReference().toProjection(),
-            projectionFactory.entity().toProjection()
+            projectionFactory.entity().toProjection(),
+            projectionFactory.score().toProjection()
```
[the remark this answers is not in the document]
Comment (fax4ever) [reply query/src/main/java/org/infinispan/query/dsl/embedded/impl/SearchQueryBuilder.java:75]: @yrodiere I wrongly thought that the score came for free, so I introduced it in all cases. Thanks for sharing it, I'm going to change the code to add it only if required by the current projections.

File: query/src/test/java/org/infinispan/query/projection/ScoreProjectionTest.java (hunk 2 above)
Comment (karesti) [reply query/src/test/java/org/infinispan/query/projection/ScoreProjectionTest.java:81]: only 4 test pass now

```[ERROR] dev.langchain4j.store.embedding.infinispan.InfinispanEmbeddingStoreIT.should_return_correct_score -- Time elapsed: 0.430 s <<< FAILURE!
java.lang.AssertionError: 

Expecting actual:
  0.7332834601402283
to be close to:
  0.9090677379700546
by less than 1% but difference was 19.336763421211277%.
(a difference of exactly 1% being considered valid)
	at dev.langchain4j.store.embedding.EmbeddingStoreWithoutMetadataIT.should_return_correct_score(EmbeddingStoreWithoutMetadataIT.java:237)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)

[ERROR] dev.langchain4j.store.embedding.infinispan.InfinispanEmbeddingStoreIT.should_find_with_min_score -- Time elapsed: 0.155 s <<< FAILURE!
java.lang.AssertionError: 

Expecting actual:
  0.7332834601402283
to be close to:
  0.9090677379700546
by less than 1% but difference was 19.336763421211277%.
(a difference of exactly 1% being considered valid)
	at dev.langchain4j.store.embedding.EmbeddingStoreWithoutMetadataIT.should_find_with_min_score(EmbeddingStoreWithoutMetadataIT.java:188)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)

[ERROR] dev.langchain4j.store.embedding.infinispan.InfinispanEmbeddingStoreIT.should_add_multiple_embeddings_with_segments -- Time elapsed: 0.152 s <<< FAILURE!
java.lang.AssertionError: 

Expecting actual:
  0.7332834601402283
to be close to:
  0.9090677379700546
by less than 1% but difference was 19.336763421211277%.
(a difference of exactly 1% being considered valid)
	at dev.langchain4j.store.embedding.EmbeddingStoreWithoutMetadataIT.should_add_multiple_embeddings_with_segments(EmbeddingStoreWithoutMetadataIT.java:160)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)

[ERROR] dev.langchain4j.store.embedding.infinispan.InfinispanEmbeddingStoreIT.should_add_multiple_embeddings -- Time elapsed: 0.145 s <<< FAILURE!
java.lang.AssertionError: 

Expecting actual:
  0.7332834601402283
to be close to:
  0.9090677379700546
by less than 1% but difference was 19.336763421211277%.
(a difference of exactly 1% being considered valid)
	at dev.langchain4j.store.embedding.EmbeddingStoreWithoutMetadataIT.should_add_multiple_embeddings(EmbeddingStoreWithoutMetadataIT.java:121)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)

[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   InfinispanEmbeddingStoreIT>EmbeddingStoreWithoutMetadataIT.should_add_multiple_embeddings:121 
Expecting actual:
  0.7332834601402283
to be close to:
  0.9090677379700546
by less than 1% but difference was 19.336763421211277%.
(a difference of exactly 1% being considered valid)
[ERROR]   InfinispanEmbeddingStoreIT>EmbeddingStoreWithoutMetadataIT.should_add_multiple_embeddings_with_segments:160 
Expecting actual:
  0.7332834601402283
to be close to:
  0.9090677379700546
by less than 1% but difference was 19.336763421211277%.
(a difference of exactly 1% being considered valid)
[ERROR]   InfinispanEmbeddingStoreIT>EmbeddingStoreWithoutMetadataIT.should_find_with_min_score:188 
Expecting actual:
  0.7332834601402283
to be close to:
  0.9090677379700546
by less than 1% but difference was 19.336763421211277%.
(a difference of exactly 1% being considered valid)
[ERROR]   InfinispanEmbeddingStoreIT>EmbeddingStoreWithoutMetadataIT.should_return_correct_score:237 
Expecting actual:
  0.7332834601402283
to be close to:
  0.9090677379700546
by less than 1% but difference was 19.336763421211277%.
(a difference of exactly 1% being considered valid)
[ERROR] Errors: 
[ERROR]   InfinispanEmbeddingStoreIT>EmbeddingStoreWithoutMetadataIT.beforeEach:24->clearStore:45 » Transport io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: localhost/[0:0:0:0:0:0:0:1]:11222
[INFO] 
[ERROR] Tests run: 8, Failures: 4, Errors: 1, Skipped: 0
[INFO] 
```

**Review by karesti — CHANGES_REQUESTED** [review]
test don't pass with the current implem

**Comment by fax4ever** [comment]
> test don't pass with the current implem

Are you using `Query<Object[]>`? Let's check it on Zulip chat

File: query/src/test/java/org/infinispan/query/projection/ScoreProjectionTest.java (hunk 2 above)
Comment (fax4ever) [reply query/src/test/java/org/infinispan/query/projection/ScoreProjectionTest.java:81]: Hibernate Search returns a float, as Lucene does. I don't know if it makes sense to cast it to double from our side...

**Comment by fax4ever** [comment]
@marko-bekhta @yrodiere thanks for the review I'm going to apply the changes

File: query/src/test/java/org/infinispan/query/projection/ScoreProjectionTest.java (hunk 2 above)
Comment (yrodiere) [reply query/src/test/java/org/infinispan/query/projection/ScoreProjectionTest.java:81]: I'm guessing Katia can do that in the Langchain4j SPI implementation?

Anyway, the problem here seems to be deeper: the test expects a certain value for the score, and that's not what it gets. No idea why... ? How was that expected score value determined?

**Review by karesti — APPROVED** [review]
It worked, all test pass in langchain integration

File: query/src/test/java/org/infinispan/query/projection/ScoreProjectionTest.java (hunk 2 above)
Comment (fax4ever) [reply query/src/test/java/org/infinispan/query/projection/ScoreProjectionTest.java:81]: @yrodiere IIUC we only need to change the distance function to cosine and after that all tests pass

**Comment by fax4ever** [comment]
@karesti thanks for checking it directly with Langchain!

**Review by yrodiere — APPROVED** [review]

**Comment by fax4ever** [comment]
Thank you @tristantarrant

**Pull request merged**